exploration/0003-formatted-parts.md

```diff
@@ -0,0 +1,187 @@
+# Formatted Parts
+
+Status: **Proposed**
+
+<details>
+	<summary>Metadata</summary>
+	<dl>
+		<dt>Contributors</dt>
+		<dd>@eemeli</dd>
+		<dt>First proposed</dt>
+		<dd>2023-08-29</dd>
+		<dt>Pull Request</dt>
+		<dd>#000</dd>
+	</dl>
+</details>
+
+## Objective
+
+Messages often include structures and values
+that are not necessarily best represented by a plain concatenated string.
+
+It would be useful to define a formatted-parts target for MessageFormat 2.
+
+## Background
+
+Past examples have shown us that if we don't provide a formatter to parts,
+the string output will be re-parsed and re-processed by users.
+
+## Use-Cases
+
+- Markup elements
+- Non-string values
+- Message post-processors
+
+## Requirements
+
+- Define an iterable sequence of formatted part objects.
+- Include metadata for each part, such as type, source, direction, and locale.
+- Allow the representation of non-string values.
+- Allow the representation of values that consist of an iterable sequence of formatted parts.
+- Be able to represent each resolved value of a pattern with any number of formatted parts, including none.
+- Define the formatted parts in a manner that allows synonymous but appropriate implementations in different programming languages.
+
+## Constraints
+
+- The JS Intl formatters already include formatted-parts representations for each supported data type.
+  The JS implementation of the MF2 formatted-parts representation should be able to match their structure,
+  at least as far as that's possible and appropriate.
+
+## Proposed Design
+
+The formatted-parts API is included in the spec as an optional but recommended formatting target.
+
+The shape of the formatted-parts output is defined in a manner similar to the data model,
+which includes TypeScript, JSON Schema, and XML DTD definitions of the same data structure.
+
+At the top level, the formatted-parts result is an iterable sequence of parts.
+Parts corresponding to each _text_ can be simpler than those of _expressions_,
+as they do not have a `source` other than their `value`,
+or set any of the other possible metadata fields.
+
+```ts
+type MessageParts = Iterable<
+  MessageTextPart | MessageExpressionPart | MessageBiDiIsolationPart
+>;
+
+interface MessageTextPart {
+  type: "text";
+  value: string;
+}
+```
+
+For MessageExpressionPart, the `source` corresponds to the expression's fallback value.
+The parts' `dir` and `locale` may be defined if overridden by an expression attribute or set otherwise.
+Each part should have at most one of `value` or `parts` defined;
+some may have none.
```

Comment:
Does this make sense? Or would it be better to say:

```suggestion
Each part MUST have either a `value` or `parts` defined. 
A part MAY have a `value` that is the empty string.
A part MAY have a `parts` that is an empty list.
```

Reply:
The suggestion would make the proposed MessageFallbackPart invalid, as it does not include either `value` or `parts`. It's conceivable for other parts to also exist which do not include either, such as open/close expressions without an annotation.

Reply:
Would an empty `value` or empty `parts` satisfy that? Or a fallback could have a string expression? Empty strings don't result in the erroneous emission of the string `null` 😉 

I understand that it would "break" the current definition: we should decide what the shapes should be and make consistent.

Reply:
For fallback when formatting to a string the `{...}` make sense as a visual indicator, but for a formatted-parts consumer some different representation could be better. So including an explicit `value` or `parts` would be misleading.

For open/close, it doesn't make sense to define their explicit parts shapes in this spec, but for JS I have them as:

```ts
interface MessageMarkupPart {
  type: 'open' | 'close';
  source: string;
  name: string;
  value?: unknown;
  options: { [key: string]: unknown };
}
```

There, the `value` would be `'b'` for `{b +html}`, but it would not be set for `{+html.b}`. Setting it to an empty string would be misleading, as `{+foo}` and `{|| +foo}` could easily mean different things.

Reply:
In https://github.com/unicode-org/message-format-wg/pull/463#discussion_r1406427143 I'm suggesting to define separate interfaces for single-valued and multi-valued parts. This could extend to fallback parts and markup, as well.